system_metrics_collector/test/topic_statistics_collector/test_received_message_age.cpp

```diff
@@ -0,0 +1,157 @@
+#include <string>
+
+#include "topic_statistics_collector/received_message_age.hpp"
+
+#include "rclcpp/clock.hpp"
+#include "rcl/time.h"
+#include "sensor_msgs/msg/imu.hpp"
+#include "std_msgs/msg/string.hpp"
+
+
+namespace
+{
+constexpr const std::chrono::seconds kDefaultDurationSeconds{1};
+constexpr const double kExpectedAverageMilliseconds{2000.0};
+constexpr const double kExpectedMinMilliseconds{1000.0};
+constexpr const double kExpectedMaxMilliseconds{3000.0};
+constexpr const double kExpectedStandardDeviation{816};
+const rclcpp::Time kDefaultSystemTime{0, 0, RCL_SYSTEM_TIME};
+constexpr const int kDefaultTimesToTest{10};
+}  // namespace
+
+class TestReceivedMessageAgeCollector
+  : public topic_statistics_collector::ReceivedMessageAgeCollector<
+    sensor_msgs::msg::Imu>
+{
+public:
+  TestReceivedMessageAgeCollector()
+  {
+    fake_now_ = ReceivedMessageAgeCollector::GetCurrentTime().nanoseconds();
+  }
+  virtual ~TestReceivedMessageAgeCollector() = default;
+
+  /**
+   * Overridden in order to mock the clock for measurement testing.
+   * @return a fixed time point
+   */
+  rclcpp::Time GetCurrentTime() override
+  {
+    return rclcpp::Time{fake_now_, RCL_SYSTEM_TIME};
+  }
+
+  /**
+   * Advance time by a specified duration, in seconds.
+   * @param seconds duration which to advance time
+   */
+  void AdvanceTime(std::chrono::nanoseconds nanos)
+  {
+    fake_now_ += nanos.count();
+  }
+
+  int64_t fake_now_;
+};
+
+sensor_msgs::msg::Imu GetImuMessageWithHeader(const int64_t timestamp)
+{
+  auto message = sensor_msgs::msg::Imu();
+  message.header = std_msgs::msg::Header();
+  message.header.stamp = rclcpp::Time(timestamp);
+  return message;
+}
+
+std_msgs::msg::String GetStringMessageWithoutHeader()
+{
+  auto message = std_msgs::msg::String();
+  message.data = "Any message with no header";
+  return message;
+}
+
+TEST(ReceivedMessageAgeTest, TestOnlyMessagesWithHeaderGetSampled) {
+  rclcpp::Clock clock{RCL_SYSTEM_TIME};
+  topic_statistics_collector::ReceivedMessageAgeCollector<std_msgs::msg::String>
+  string_msg_collector{clock};
+
+  const auto string_msg = GetStringMessageWithoutHeader();
+  moving_average_statistics::StatisticData stats;
+
+  for (int i = 0; i < kDefaultTimesToTest; ++i) {
+    string_msg_collector.OnMessageReceived(string_msg);
+    stats = string_msg_collector.GetStatisticsResults();
+    EXPECT_EQ(0, stats.sample_count) << "Expect 0 samples to be collected";
+  }
+
+  topic_statistics_collector::ReceivedMessageAgeCollector<sensor_msgs::msg::Imu>
+  imu_msg_collector{clock};
+  const auto imu_msg = GetImuMessageWithHeader(clock.now().nanoseconds());
+
+  for (int i = 0; i < kDefaultTimesToTest; ++i) {
+    imu_msg_collector.OnMessageReceived(imu_msg);
+    stats = imu_msg_collector.GetStatisticsResults();
+    EXPECT_EQ(i + 1, stats.sample_count) << "Expect " << i + 1 << " samples to be collected";
+  }
+}
+
+TEST(ReceivedMessageAgeTest, TestMeasurementOnlyMadeForPositiveHeaderValue) {
```

Comment:
Is a negative header value possible?

Reply:
Depending on how it is initialized using `rclcpp::Time`, `header` could have negative values, or not (see [here](https://github.com/ros2/rclcpp/blob/3e08d9e43f67acdbdf1bef66f2b1ebf410af594d/rclcpp/src/rclcpp/time.cpp#L49-L64)).

The intent here is to filter out _uninitialized_ `headers` from the measurement because the metric doesn't make sense if incoming message timestamp is 0. I will rename the test, the intent isn't clear from the name.